lib/manager/actions/deployments.js

```diff
@@ -259,6 +260,52 @@ export function updateVersionForFeedSource (
   }
 }
 
+/**
+ * Uploads a csv file to the server, which uploads it to s3. Associates the resulting URL with
+ * a deployment
+ * @param {*} deployment                    The deployment to add the csv file to
+ * @param {*} file                          The csv file to upload
+ * @param {*} fileToDeleteOnSuccesfulUpload A URL to delete from the deployment on a successful upload
+ * @returns                                 false if something fails, otherwise the deployment update
+ */
+export async function uploadPeliasWebhookCsvFile (deployment: Deployment, file: File, fileToDeleteOnSuccesfulUpload?: string | null) {
+  return async function (dispatch: dispatchFn, getState: getStateFn) {
+    const { user } = getState()
+    if (!user || !user.token) {
+      window.alert('You are not logged in. Please log in again.')
+      return false
+    }
+    const { id } = deployment
+    const url = `${DEPLOYMENT_URL}/${id}/upload`
+    let peliasCsvFiles = deployment.peliasCsvFiles || []
+
+    const formData = new FormData()
+    formData.append('file', file)
+
+    // Have to use custom fetch because Spark is very particular about which
+    // headers must accompany a multipart file request
+    const s3UploadResponse = await fetch(url, {
+      method: 'post',
+      // $FlowFixMe user is checked above, but Flow doesn't recognize it
+      headers: getHeaders(user.token, false),
+      body: formData
+    })
+    if (s3UploadResponse.status >= 400) {
+      const errorJson = await s3UploadResponse.json()
+      errorJson && window.alert(errorJson.message || 'Could not upload CSV file')
+      return false
+    }
+    const s3UploadUrl = await s3UploadResponse.text()
+    // If a file is set to delete on successful upload, delete it here
+    if (fileToDeleteOnSuccesfulUpload) {
+      peliasCsvFiles = peliasCsvFiles.filter(csv => csv !== fileToDeleteOnSuccesfulUpload)
+    }
+
+    await dispatch(updateDeployment(deployment, { 'peliasCsvFiles': [ ...peliasCsvFiles, s3UploadUrl ] }))
```

Comment:
This flow seems strange to me: shouldn't we just let the server handle updating this field and deleting the file?

Reply:
Also, when trying to upload a csv, I hit a null pointer exception here:
https://github.com/ibi-group/datatools-server/blob/71ebcc663ce43d48264de6469368d9d221bf876d/src/main/java/com/conveyal/datatools/manager/controllers/api/DeploymentController.java#L333

(Probably due to `#peliasCsvFiles` field being null.

Reply:
Also, the errors encountered on the front end do not seem to be handled correctly. It might be wise to surround a fair amount of this in a try/catch.

Reply:
Have made some changes that should make errors like this easier to debug: 30cb8b3b6d8a3dad68754e7c30388a0c7f13cea9

Have also moved the csv replace action to the server in 45c84c7bb932830ffcaa73e4f72358565103dc43